core/vcpu.c

```diff
@@ -4311,6 +4316,66 @@ static void vcpu_state_dump(struct vcpu_t *vcpu)
             vcpu->state->_idt.base,
             vcpu->state->_idt.limit,
             (uint32_t)vcpu->state->_efer);
+
+    // Dump FPU state
+    hax_log(HAX_LOGW,
+            "FX layout:\n"
+            "fcw:    %08x  fsw:    %08x  ftw: %08x  res1: %08x  fop: %08x\n"
+            "fpu_ip: %08llx  fpu_dp: %08llx\n"
+            "mxcsr:  %08x  mxcsr_mask: %08x\n",
+            gfx->fcw, gfx->fsw, gfx->ftw, gfx->res1, gfx->fop, gfx->fpu_ip,
+            gfx->fpu_dp, gfx->mxcsr, gfx->mxcsr_mask);
+
+    for (i = 0; i < 16; i++) {
+        hax_log(HAX_LOGW, "st_mm[%d] %02x%02x%02x%02x%02x%02x%02x%02x\n", i,
+                gfx->st_mm[0][i], gfx->st_mm[1][i], gfx->st_mm[2][i],
```

Comment:
Could you explain why you output the array by column? According to the 2-dimensional array in memory layout, do you think it is more reasonable as below format?

    for (i = 0; i < 8; i++) {
        hax_log(HAX_LOGW, "st_mm[%d]: %02x%02x%02x%02x %02x%02x%02x%02x "
                "%02x%02x%02x%02x %02x%02x%02x%02x\n", i, gfx->st_mm[i][0],
                gfx->st_mm[i][1], gfx->st_mm[i][2], gfx->st_mm[i][3],
                gfx->st_mm[i][4], gfx->st_mm[i][5], gfx->st_mm[i][6],
                gfx->st_mm[i][7], gfx->st_mm[i][8], gfx->st_mm[i][9],
                gfx->st_mm[i][10], gfx->st_mm[i][11], gfx->st_mm[i][12],
                gfx->st_mm[i][13], gfx->st_mm[i][14], gfx->st_mm[i][15]);
    }

Reply:
I thought st_mm are registers of 8 byte size. Can you give me Intel docs link where these are described?

Reply:
According to the definition in SDM Vol. 2A 3.1.1.3, st_mm are 8-byte width. I am not sure why the definition of the array here is 8 by 16. But from the existing code, see vcpu_get_fpu() in core/vcpu.c, the nested loop is raws first and then columns. Anyway, I think the 16 bytes within each raw is memory consecutive. If you are sure your implementation is reasonable, I will also accept the approach as I cannot find any exact document link to confirm that. Of course, to make the format consistent with others is okay.

SDM definition:

ST(i) — The ith element from the top of the FPU register stack (i ← 0 through 7).
mm — An MMX register. The 64-bit MMX registers are: MM0 through MM7.

Reply:
Table 3-47 in sdm-vol-2abcd - these are saved by fxrstor as 10 bytes width, 6 bytes reserved. Rewritten the code for this and xmms.